xlators/storage/posix/src/posix-helpers.c

```diff
@@ -1789,9 +1816,65 @@ posix_gfid_heal(xlator_t *this, const char *path, loc_t *loc, dict_t *xattr_req)
             }
         }
     }
+    /* posix_gfid_set populate a xattr trusted.glink on the real_path
+       to make sure gfid/link has been correctly set duing any entry creation.
+       So at the time of fresh lookup there is no need to verify_handle because
+       it consumes time during lookup while huge number of files(1M)
+       are available on the backend.
+    */
 
-    (void)posix_gfid_set(this, path, loc, xattr_req, GF_CLIENT_PID_MAX, &ret);
-    return 0;
+    if (size == 16) {
+        ret = sys_lgetxattr(path, GFID_LINK_CREATED, gfid_link, 1);
+        if (ret == 1) {
+            return 0;
+        } else {
+            goto verify_handle;
+        }
+    }
+
+    ret = dict_get_gfuuid(xattr_req, "gfid-req", &uuid_req);
+    if (ret) {
+        gf_msg_debug(this->name, 0, "failed to get the gfid from dict for %s",
+                     loc->path);
+        ret = -1;
+        goto out;
+    }
+    if (gf_uuid_is_null(uuid_req)) {
+        gf_msg(this->name, GF_LOG_ERROR, EINVAL, P_MSG_NULL_GFID,
+               "gfid is null for %s", loc ? loc->path : "");
+        ret = -1;
+        goto out;
+    }
+
+    ret = sys_lsetxattr(path, GFID_XATTR_KEY, uuid_req, 16, XATTR_CREATE);
```

Comment:
For files this adds extra unnecessary syscalls. For directories in rename, the links are deleted, so the xattr goes stale. I am a bit busy this week with something else. So may respond very late for any comments.